src/safemode_ui.cpp

```diff
@@ -0,0 +1,741 @@
+    WINDOW_PTR wptr( w );
+
+    draw_border( w_border, BORDER_COLOR, custom_name );
+
+    mvwputch( w_border, 3,  0, c_ltgray, LINE_XXXO ); // |-
+    mvwputch( w_border, 3, 79, c_ltgray, LINE_XOXX ); // -|
+
+    for( auto &mapLine : mapPos ) {
+        mvwputch( w_border, FULL_SCREEN_HEIGHT - 1, mapLine.second + 1, c_ltgray,
+                  LINE_XXOX ); // _|_
+    }
+
+    wrefresh( w_border );
+
+    int tmpx = 0;
+    tmpx += shortcut_print( w_header, 0, tmpx, c_white, c_ltgreen, _( "<A>dd" ) ) + 2;
+    tmpx += shortcut_print( w_header, 0, tmpx, c_white, c_ltgreen, _( "<R>emove" ) ) + 2;
+    tmpx += shortcut_print( w_header, 0, tmpx, c_white, c_ltgreen, _( "<C>opy" ) ) + 2;
+    tmpx += shortcut_print( w_header, 0, tmpx, c_white, c_ltgreen, _( "<M>ove" ) ) + 2;
+    tmpx += shortcut_print( w_header, 0, tmpx, c_white, c_ltgreen, _( "<E>nable" ) ) + 2;
+    tmpx += shortcut_print( w_header, 0, tmpx, c_white, c_ltgreen, _( "<D>isable" ) ) + 2;
+    shortcut_print( w_header, 0, tmpx, c_white, c_ltgreen, _( "<T>est" ) );
+    tmpx = 0;
+    tmpx += shortcut_print( w_header, 1, tmpx, c_white, c_ltgreen,
+                            _( "<+-> Move up/down" ) ) + 2;
+    tmpx += shortcut_print( w_header, 1, tmpx, c_white, c_ltgreen, _( "<Enter>-Edit" ) ) + 2;
+    shortcut_print( w_header, 1, tmpx, c_white, c_ltgreen, _( "<Tab>-Switch Page" ) );
+
+    for( int i = 0; i < 78; i++ ) {
+        mvwputch( w_header, 2, i, c_ltgray, LINE_OXOX ); // Draw line under header
+    }
+
+    for( auto &mapLine : mapPos ) {
+        mvwputch( w_header, 2, mapLine.second, c_ltgray, LINE_OXXX );
+        mvwputch( w_header, 3, mapLine.second, c_ltgray, LINE_XOXO );
+    }
+
+    mvwprintz( w_header, 3, 1, c_white, "#" );
+    mvwprintz( w_header, 3, mapPos[col_Rule] + 4, c_white,
+               _( "Rules" ) ); //Creaturename with wildcard or NPC
+    mvwprintz( w_header, 3, mapPos[col_Attitude] + 2, c_white,
+               _( "Attitude" ) ); //Hostile/Neutral/Friendly
+    mvwprintz( w_header, 3, mapPos[col_Dist] + 2, c_white, _( "Dist" ) ); //Proximity distance
+    mvwprintz( w_header, 3, mapPos[col_IE] + 2, c_white, _( "I/E" ) ); //Include or Exclude rule
+
+    wrefresh( w_header );
+
+    int iTab = GLOBAL_TAB;
+    int iLine = 0;
+    int iColumn = 0;
+    int iStartPos = 0;
+    bool bStuffChanged = false;
+    input_context ctxt( "SAFEMODE" );
+    ctxt.register_cardinal();
+    ctxt.register_action( "CONFIRM" );
+    ctxt.register_action( "QUIT" );
+    ctxt.register_action( "NEXT_TAB" );
+    ctxt.register_action( "PREV_TAB" );
+    ctxt.register_action( "ADD_RULE" );
+    ctxt.register_action( "REMOVE_RULE" );
+    ctxt.register_action( "COPY_RULE" );
+    ctxt.register_action( "ENABLE_RULE" );
+    ctxt.register_action( "DISABLE_RULE" );
+    ctxt.register_action( "MOVE_RULE_UP" );
+    ctxt.register_action( "MOVE_RULE_DOWN" );
+    ctxt.register_action( "TEST_RULE" );
+    ctxt.register_action( "HELP_KEYBINDINGS" );
```

Comment:
Side note here:
We should probably have a `register_action` overload that takes a container argument.